### PR TITLE
Optimize some reduction operators on CPU BFloat16

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58297 add BFloat16 support for UpSample on CPU
* #56904 optimize transpose copy for float32 and bfloat16 on CPU
* #56903 add BFloat16 support for MaxPool2d on CPU
* #56902 add BFloat16 support for AdaptiveAvgPool2d on CPU
* #56372 add BFloat16 support for bernoulli and Dropout on CPU
* #55588 add bf16 support for bucketize
* #55221 optimize BFloat16 elemwise operators CPU: sigmoid, sigmoid_backward, tanh_backward, addcmul, addcdiv
* #55217 SumKernel (BFloat16): use float as accumulation type
* #55210 add BFloat16 support for LayerNorm CPU
* **#55202 Optimize some reduction operators on CPU BFloat16**

Differential Revision: [D28836790](https://our.internmc.facebook.com/intern/diff/D28836790)

### CODE REVIEW

**Comment by mingfeima** [comment]
This PR aims at enabling or optimizing the following reduction operators on CPU BFloat16:
- softmax (don't support BFloat16 previously)
- log_softmax
- max (reduce_all)
- min (reduce_all)
- _aminmax (reduce_all; don't support BFloat16 previously)

Note that previously some operators e.g. `log_softmax` already have BFloat16 support, but the perf is not good.
This PR specializes all the `map<>()` from **vec256/functional.h** and also corresponding test cases are added in **vec256_test_all_types.cpp**.


Since we have already specializes all members for Vec256<scalar_t> with {scalar_t=BFloat16} in vec256_bfloat16.h, so the following example would run smoothing on BFloat16:
```C++
using Vec = Vec256<BFloat16>;
Vec one = Vec(BFloat16(1));
vec256::map([](Vec x) { return one / (one + x.exp()); }, y_ptr, x_ptr, N);
```

The current impl will end up with 3 pairs of dtype conversion, each for ".exp()", "+" and "/" respectively.
The new impl will only need dtype conversion for input and output. Benefits:

- better performance since we have less dtype conversion;
- less rounding error since immediate results are kept in fp32;
- accumulation done on data type of fp32.

I am going to use this stack to push more BFloat16 CPU optimizations, following the same manner as propose in this PR:
- input load: bf16->fp32; output store: fp32->bf16
- all immediate operations (including accumulation) will use fp32

**Comment by mingfeima** [comment]
Since this PR is not related to parallelization feature, only single core perf is tested:
NB: if the operator doesn't have BFloat16 support previously, the `before` perf refers to simple impl as following:
```diff
-   AT_DISPATCH_ALL_TYPES(input.scalar_type(), "_aminmax_all_all", [&] {
+   AT_DISPATCH_ALL_TYPES_AND(kBFloat16, input.scalar_type(), "_aminmax_all_all", [&] {
```

- performance update on **avx512** machine: Xeon(R) Gold 6248 CPU @ 2.50GHz
```bash
before: softmax: 128x1024: fp32: 151.457 us; bf16: 362.440 us
after:  softmax: 128x1024: fp32: 151.757 us; bf16: 194.105 us

before: log_softmax: 128x1024: fp32: 157.474 us; bf16: 411.537 us
after:  log_softmax: 128x1024: fp32: 152.229 us; bf16: 163.657 us

before: max: 128x1024: fp32: 23.714 us; bf16: 63.077 us
after:  max: 128x1024: fp32: 24.523 us; bf16: 17.484 us

before: min: 128x1024: fp32: 23.707 us; bf16: 63.198 us
after:  min: 128x1024: fp32: 24.067 us; bf16: 17.498 us

before: _aminmax: 128x1024: fp32: 25.156 us; bf16: 69.929 us
after:  _aminmax: 128x1024: fp32: 25.272 us; bf16: 19.487 us
```

- performance update on **avx2** machine: Xeon(R) CPU E5-2680 v3 @ 2.50GHz
```bash
before: softmax: 128x1024: fp32: 229.048 us; bf16: 493.946 us
after:  softmax: 128x1024: fp32: 248.060 us; bf16: 293.789 us

before: log_softmax: 128x1024: fp32: 274.467 us; bf16: 673.550 us
after:  log_softmax: 128x1024: fp32: 251.243 us; bf16: 241.080 us

before: max: 128x1024: fp32: 36.288 us; bf16: 88.725 us
after:  max: 128x1024: fp32: 32.229 us; bf16: 23.905 us

before: min: 128x1024: fp32: 32.179 us; bf16: 87.669 us
after:  min: 128x1024: fp32: 30.829 us; bf16: 22.847 us

before: _aminmax: 128x1024: fp32: 33.098 us; bf16: 105.431 us
after:  _aminmax: 128x1024: fp32: 30.785 us; bf16: 28.113 us
```


Notes:
- With this PR, BFloat16 Softmax is still slower than float32, because lack of native dtype conversion intrinsics (currently fp32/bf16 conversion uses emulated method), on Sapphire Rapids BFloat16 is faster.

**Comment by mingfeima** [comment]
@VitalyFedyunin Could you please review this stack? We are trying to optimize CPU BFloat16 path performance.

**Comment by mdschatz** [comment]
@mdschatz has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28690413).

**Comment by mdschatz** [comment]
@mdschatz has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28690481).

**Comment by mdschatz** [comment]
@mdschatz has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28695688).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836790).

**Comment by VitalyFedyunin** [comment]
Hi! Please rebase to adapt to more general vectorization as we are in process of introducing AVX512 support.

**Comment by mingfeima** [comment]
@VitalyFedyunin Hi, this stack has been rebased!

File: aten/src/ATen/cpu/vec/vec256/functional.h [hunk 1]
```diff
@@ -231,4 +231,444 @@ inline void map3(
   }
 }
 
+// BFloat16 specification
```
Comment (VitalyFedyunin) [inline aten/src/ATen/cpu/vec/vec256/functional.h:234]: Please move bfloat16 related code into the `vec256_bfloat16.h`

Comment (VitalyFedyunin) [reply aten/src/ATen/cpu/vec/vec256/functional.h:234]: If necessary (circular deps) create another file, something like _mixed.h

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 2]
```diff
@@ -45,11 +45,17 @@ void cat_serial_kernel_impl(Tensor& result, TensorList tensors, int64_t dim) {
           result_ptr[k] = input_ptr[k];
         }
       } else {
-        vec::map(
-            [](Vec x) { return x; },
-            result_ptr,
-            input_ptr,
-            local_inner);
+        int64_t d = 0;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/CatKernel.cpp:48]: Seems to be, unrelated change, can you put it as separate PR

File: aten/src/ATen/test/vec_test_all_types.cpp [hunk 3]
```diff
@@ -1297,6 +1301,128 @@ namespace {
             },
             test_case);
     }
+    TYPED_TEST(FunctionalBF16Tests, Reduce) {
+      using vec = TypeParam;
+      // Can't use ValueType<TypeParam> here:
+      // Vectorized<BFloat16>::value_type returns uint16_t on AVX2
```
Comment (VitalyFedyunin) [inline aten/src/ATen/test/vec_test_all_types.cpp:1307]: Can you please file GH issue for it and reference this piece for removal/update.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 4]
```diff
@@ -55,8 +55,9 @@ template <typename T>
 void VmlLog(int64_t N, const T* X, T* Y) {
   constexpr int64_t K = Vectorized<T>::size();
   at::parallel_for(0, N, K, [=](int64_t begin, int64_t end) {
+    using VT = vec::vec_scalar_t<T>;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:58]: Looks unrelated to reductions

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Some code moves required, otherwise looks fine

File: aten/src/ATen/cpu/vec/vec256/functional.h (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/cpu/vec/vec256/functional.h]: Done. Separate bfloat16 part into `functional_bfloat16.h`

File: aten/src/ATen/native/cpu/CatKernel.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/CatKernel.cpp]: Indeed this is unrelated, but since i changed functional specification for bfloat16. The behavior of `map<BFloat16>` changed:
* (before): `map<BFloat16>` would require the vec type to be `Vectorized<BFloat16>`
* (after): `map<BFloat16>` would require the vec type to be `Vectorized<float>`

Using `Vectorized<float>` is aiming at reducing the unnecessary dtype change between fp32 and bf16. AKA. only input and output data will do the dtype change and only fp32 is used in the computation for better perf and less rounding error.

So if is not changed, there would be compilation error.

However, after the change of `map<BFloat16>` behavior, it is no longer suitable for `cat` which is only a simple memory copy (no computation).

if we stick to use `map<BFloat16>`, it is going to be:
```
1. load as bf16 vec (length = 16)
2. convert into 2 * fp32 vec (length = 2 * 8)
3. convert back to bf16 vec (length = 16)
4. store to dst address
```

This is not our intention. A simple load/store will be enough here. So for the simple memcpy case it is no longer suitable to use `map<>` for BFloat16. I haven't notice other similar circumstances (using `map<>` for memcpy on BFloat16), will modify if met any.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 4 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:58]: similar to above, since `map<BFloat16>` would require vec type to be `Vectorized<float>`, this would lead to compilation error if un-changed.

**Comment by mingfeima** [comment]
@VitalyFedyunin, updated! please check.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836790).

**Comment by mdschatz** [comment]
@mingfeima, I've noticed that including `vec256_float.h` is needed in `vec256_bfloat16.h` in order to get compilation to proceed; without it, I am getting a lot of __m256 type-cast errors.  Should that line be included in this PR or not?  @jgong5 Can you elaborate on this?

**Comment by jgong5** [comment]
@mdschatz The compilation error was due to an incorrect header file inclusion in IPEX (should include `ATen/cpu/vec/vec.h` instead of `vec256_bfloat16.h` directly, and we have fixed it there. So, no worries!

**Comment by mingfeima** [comment]
> @mdschatz The compilation error was due to an incorrect header file inclusion in IPEX (should include `ATen/cpu/vec/vec.h` instead of `vec256_bfloat16.h` directly, and we have fixed it there. So, no worries!

right, you should include `<ATen/cpu/vec/vec.h>` instead of directly include 'vec256_xxx.h'
